compiler/src/org.graalvm.compiler.replacements/src/org/graalvm/compiler/replacements/DefaultJavaLoweringProvider.java

```diff
@@ -972,10 +972,12 @@ private BarrierType guessStoreBarrierType(ValueNode object, ValueNode value) {
             ResolvedJavaType type = StampTool.typeOrNull(object);
             // Array types must use a precise barrier, so if the type is unknown or is a supertype
             // of Object[] then treat it as an array.
-            if (type == null || type.isArray() || type.isAssignableFrom(objectArrayType)) {
-                return BarrierType.PRECISE;
+            if (type == null || type.isAssignableFrom(objectArrayType)) {
```

Comment:
I think you need to rearrange these tests.  type.isArray() must be tested before type.isAssignableFrom(objectArrayType) to find the ARRAY case.

Reply:
Latest push should fix it.